pkg/prom/wal/wal.go

```diff
@@ -90,6 +94,12 @@ type Storage struct {
 	storage.Queryable
 	storage.ChunkQueryable
 
+	// Operations against the WAL must be protected by a mutex so it doesn't get closed
+	// in the middle of an operation. If the WAL is closed, operations that change the WAL
+	// must fail.
+	walMtx    sync.Mutex
+	walClosed chan bool
```

Comment:
The change looks good, but I'm missing why this needs to be a channel instead of simply a flag (`walClosed bool`).

`Storage.Close()` takes the mutex, so it prevents the other functions from advancing while the close operation is still in progress, and `Storage.isWALClosed()` is always called with the mutex held and it never blocks, so closing the channel in `Storage.Close()` doesn't unblock the other operations.

Taking a step back, are the other operations besides `Close()` concurrent-safe? I'm asking because the way I read the code, beyond fixing the bug this change is also serializing those. Would it be an option to use a `sync.RWMutex` that `Close()` takes for writing (excluding the other uses) and the other operations take only for reading (preventing `Close()` from advancing, but allowing the other operations to proceed concurrently)?

Reply:
Channel was a force of habit, it can totally just be a boolean and I'll change that. 👍 

> Taking a step back, are the other operations besides Close() concurrent-safe? I'm asking because the way I read the code, beyond fixing the bug this change is also serializing those. 

All other operations are concurrent safe; the WAL is designed to be accessed from multiple goroutines concurrently, with the exception of the `Close` that I missed in my usage here. 

> Would it be an option to use a sync.RWMutex that Close() takes for writing (excluding the other uses) and the other operations take only for reading (preventing Close() from advancing, but allowing the other operations to proceed concurrently)?`

I think this makes sense, but it's only going to be a marginal improvement. Most of the concurrent access is going to be in `appender.Commit` which will immediately get blocked if trying to log samples to the WAL concurrently. However, it'd still allow them to encode the records concurrently, so that's a small improvement. 

I'll make both of these changes. Thanks for catching the problems in my rushed implementation here!

Reply:
Changed, PTAL 🙏

Reply:
Thanks! Looks good!